state-commitments/optimized/history_commitment.go

```diff
@@ -0,0 +1,461 @@
+				}
+			} else {
+				if len(h.lastLeafFillers) > 0 {
+					right, err = h.computeVirtualSparseTree([]common.Hash{h.lastLeafFillers[0]}, virtual-limit/2, limit/2)
+					if err != nil {
+						return common.Hash{}, err
+					}
+				} else {
+					return common.Hash{}, errors.New("empty lastLeafFillers")
+				}
+			}
+		}
+	} else {
+		left, err = h.computeVirtualSparseTree(leaves, virtual, limit/2)
+		if err != nil {
+			return common.Hash{}, err
+		}
+		right = emptyHash
+	}
+	if _, err = h.keccak.Write(left[:]); err != nil {
+		return common.Hash{}, err
+	}
+	if _, err = h.keccak.Write(right[:]); err != nil {
+		return common.Hash{}, err
+	}
+	if _, err = h.keccak.Read(leaves[0][:]); err != nil {
+		return common.Hash{}, err
+	}
+	h.keccak.Reset()
+	return leaves[0], nil
+}
+
+func (h *HistoryCommitter) subtreeExpansion(leaves []common.Hash, virtual, limit uint64, stripped bool) (proof []common.Hash, err error) {
+	m := uint64(len(leaves))
+	if m == 0 {
+		return make([]common.Hash, 0), nil
+	}
+	if virtual == 0 {
+		for i := limit; i > 1; i /= 2 {
+			proof = append(proof, emptyHash)
+		}
+		return
+	}
+	if limit == 0 {
+		limit = nextPowerOf2(virtual)
+	}
+	if limit == virtual {
+		left, err2 := h.computeSparseTree(leaves, limit, h.lastLeafFillers)
+		if err2 != nil {
+			return nil, err2
+		}
+		if !stripped {
+			for i := limit; i > 1; i /= 2 {
+				proof = append(proof, emptyHash)
+			}
+		}
+		return append(proof, left), nil
+	}
+	if m > limit/2 {
+		left, err2 := h.computeSparseTree(leaves[:limit/2], limit/2, nil)
+		if err2 != nil {
+			return nil, err2
+		}
+		proof, err = h.subtreeExpansion(leaves[limit/2:], virtual-limit/2, limit/2, stripped)
+		if err != nil {
+			return nil, err
+		}
+		return append(proof, left), nil
+	}
+	if virtual >= limit/2 {
+		left, err2 := h.computeSparseTree(leaves, limit/2, h.lastLeafFillers)
+		if err2 != nil {
+			return nil, err2
+		}
+		// Check if h.lastLeafFillers is not empty before accessing its first element
+		if len(h.lastLeafFillers) > 0 {
+			proof, err = h.subtreeExpansion([]common.Hash{h.lastLeafFillers[0]}, virtual-limit/2, limit/2, stripped)
+			if err != nil {
+				return nil, err
+			}
+			return append(proof, left), nil
+		} else {
+			return nil, errors.New("lastLeafFillers is empty")
+		}
+	}
+	if stripped {
+		return h.subtreeExpansion(leaves, virtual, limit/2, stripped)
+	}
+	expac, err := h.subtreeExpansion(leaves, virtual, limit/2, stripped)
+	if err != nil {
+		return nil, err
+	}
+	return append(expac, emptyHash), nil
+}
+
+func (h *HistoryCommitter) proof(index uint64, leaves []common.Hash, virtual, limit uint64) (tail []common.Hash, err error) {
+	m := uint64(len(leaves))
+	if m == 0 {
+		return nil, errors.New("empty leaves slice")
+	}
+	if limit == 0 {
+		limit = nextPowerOf2(virtual)
+	}
+	if limit == 1 {
+		// Can only reach this with index == 0
+		return
+	}
+	if index >= limit/2 {
+		if m > limit/2 {
+			if limit/2 < m {
```

Comment:
This is the exact same check than the line above

Reply:
Removed.